src/storage/sftp/storage.c

```diff
@@ -37,6 +38,199 @@ struct StorageSftp
+            break;
+#endif
+
+#ifdef LIBSSH2_HOSTKEY_TYPE_ECDSA_521
+        case LIBSSH2_HOSTKEY_TYPE_ECDSA_521:
+            knownHostKeyType = LIBSSH2_KNOWNHOST_KEY_ECDSA_521;
+            break;
+#endif
+
+#ifdef LIBSSH2_HOSTKEY_TYPE_ED25519
+        case LIBSSH2_HOSTKEY_TYPE_ED25519:
+            knownHostKeyType = LIBSSH2_KNOWNHOST_KEY_ED25519;
+            break;
+#endif
+
+        default:
+            knownHostKeyType = 0;
+            break;
+    }
+
+    FUNCTION_TEST_RETURN(INT, knownHostKeyType);
+}
+
+/***********************************************************************************************************************************
+Return a match failed message based on known host check failure type
+***********************************************************************************************************************************/
+static const char *
+storageSftpKnownHostCheckpFailureMsg(const int rc)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(INT, rc);
+    FUNCTION_TEST_END();
+
+    const char *matchFailMsg;
+
+    switch (rc)
+    {
+        case LIBSSH2_KNOWNHOST_CHECK_FAILURE:
+            matchFailMsg = "LIBSSH2_KNOWNHOST_CHECK_FAILURE";
+            break;
+
+        case LIBSSH2_KNOWNHOST_CHECK_NOTFOUND:
+            matchFailMsg = "not found in known hosts files: LIBSSH2_KNOWNHOST_CHECK_NOTFOUND";
+            break;
+
+        case LIBSSH2_KNOWNHOST_CHECK_MISMATCH:
+            matchFailMsg = "mismatch in known hosts files: LIBSSH2_KNOWNHOST_CHECK_MISMATCH";
+            break;
+
+        default:
+            matchFailMsg = "experienced an unknown failure:";
+            break;
+    }
+
+    FUNCTION_TEST_RETURN_CONST(STRINGZ, matchFailMsg);
+}
+
+/***********************************************************************************************************************************
+Rewrite the user's known_hosts file with a new entry
+***********************************************************************************************************************************/
+static void
+storageSftpUpdateKnownHostsFile(StorageSftp *const this, const int hostKeyType, const String *const host, const char *const hostKey,
+        const size_t hostKeyLen)
+{
+    FUNCTION_LOG_BEGIN(logLevelDebug);
+        FUNCTION_LOG_PARAM(STORAGE_SFTP, this);
+        FUNCTION_LOG_PARAM(INT, hostKeyType);
+        FUNCTION_LOG_PARAM(STRING, host);
+        FUNCTION_LOG_PARAM(STRINGZ, hostKey);
+        FUNCTION_LOG_PARAM(SIZE, hostKeyLen);
+    FUNCTION_LOG_END();
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        LIBSSH2_KNOWNHOSTS *userKnownHostsList;
+        char *libSsh2ErrMsg;
+        int libSsh2ErrMsgLen;
+        int rc;
+
+        // Init a known host collection for the user's known_hosts file
+        const char *const userKnownHostsFile = strZ(strNewFmt("%s%s", strZ(userHome()), "/.ssh/known_hosts"));
+        userKnownHostsList = libssh2_knownhost_init(this->session);
```

Comment:
Callers to this function have already done an init, why aren't we passing down that into this function and using it..?

Reply:
The initial collection will load up the entries in either the default hosts file list or the file(s) provided by the repo-sftp-known-host values. The storageSftpUpdateKnownHostsFile() function needs to operate on only the user's known_hosts file. If we use the initial collection then we'd write out all entries from all files to the user's known_hosts when we add a new host entry.